tar/lib/src/stream.dart

```diff
@@ -0,0 +1,135 @@
+  ArgumentError.checkNotNull(source, 'source');
+  ArgumentError.checkNotNull(sparseHoles, 'sparseHoles');
+  ArgumentError.checkNotNull(size, 'size');
+
+  /// Current logical position in sparse file, including the bytes in
+  /// [partialChunk].
+  var position = 0;
+
+  /// Index of the next sparse hole in [sparseHoles] to be processed.
+  var sparseHoleIndex = 0;
+
+  /// Iterator through [source] to obtain the data bytes.
+  final iterator = StreamIterator(source);
+
+  while (position < size) {
+    /// Yield all the necessary sparse holes.
+    while (sparseHoleIndex < sparseHoles.length) {
+      final sparseHole = sparseHoles[sparseHoleIndex];
+      if (sparseHole.offset == position) {
+        for (final byte in List<int>.filled(sparseHole.length, 0)) {
+          yield byte;
+        }
+        position += sparseHole.length;
+        sparseHoleIndex++;
+      } else {
+        break;
+      }
+    }
+
+    if (position == size) break;
+
+    /// Grab a new chunk from [source].
+    if (!await iterator.moveNext()) {
+      throw StateError('sparseStream size is larger than what the stream '
+          'supports!');
+    }
+    yield iterator.current;
+    position++;
+  }
+}
+
+/// Implementation of a Stream iterator that reads a chunked stream of [List<T>].
+class ChunkedStreamIterator<T> {
+  /// Underlying iterator that iterates through the original stream.
+  final StreamIterator<T> _iterator;
+
+  /// Keeps track of the number of elements left in the current substream.
+  int _toRead = 0;
+
+  /// Keeps track of the current substream we are supporting.
+  int _substreamId = 0;
+
+  ChunkedStreamIterator(Stream<List<T>> stream)
+      : _iterator = StreamIterator(stream.expand((element) => element));
+
+  /// Returns a block of the next [size] elements.
+  ///
+  /// Returns a list with less than [size] elements if the end of stream is
+  /// encounted before [size] elements are read.
+  ///
+  /// If an error is encountered before reading [size] elements, the error
+  /// will be thrown.
+  Future<List<T>> read(int size) async {
+    /// Clears the remainder of elements if the user forgot to drain it.
+    while (_toRead > 0 && await _iterator.moveNext()) {
+      _toRead--;
+    }
+
+    final result = <T>[];
+
+    while (size-- > 0 && await _iterator.moveNext()) {
+      result.add(_iterator.current);
+    }
+
+    return result;
+  }
+
+  /// Cancels the stream iterator (and the underlying stream subscription) early.
+  ///
+  /// The [ChunkedStreamIterator] is automatically cancelled if [read] goes
+  /// reaches the end of the stream or an error.
+  ///
+  /// Users should call [cancel] to ensure that the stream is properly closed
+  /// if they need to stop listening earlier than the end of the stream.
+  ///
+  /// Returns a future if the [cancel] operation is not completed
+  /// asynchronously. Otherwise returns null.
+  Future<dynamic> cancel() => _iterator.cancel();
```

Comment:
I wuld avoid `Future<dynamic>` if there is not response the reader should care about just do: `Future<void>`